packages/@aws-cdk/aws-codepipeline/test/fake-build-action.ts

```diff
@@ -8,19 +10,33 @@ export interface FakeBuildActionProps extends codepipeline.CommonActionProps {
   extraInputs?: codepipeline.Artifact[];
 }
 
-export class FakeBuildAction extends codepipeline.Action {
+export class FakeBuildAction implements codepipeline.IAction {
+  public readonly actionName: string;
+  public readonly category = codepipeline.ActionCategory.BUILD;
+  public readonly provider = 'Fake';
+  public readonly artifactBounds =  { minInputs: 1, maxInputs: 3, minOutputs: 0, maxOutputs: 1 };
+  public readonly inputs?: codepipeline.Artifact[];
+  public readonly outputs?: codepipeline.Artifact[];
+
+  public readonly owner = undefined;
+  public readonly version = undefined;
+  public readonly runOrder = undefined;
+  public readonly region = undefined;
+  public readonly resource = undefined;
+  public readonly role = undefined;
```

Comment:
This actually gets quite tedious, and is not actually that great for future extensibility (if I add a property to `IAction`, even an optional one, I would have to update this class, and every other `IAction` implementation).

I'm wondering whether it's better to introduce a data interface for all these, and then just have on method in `IAction`, something like `getActionProperties()`, that returns it. This way, implementing the interface will be much more concise, and it will allow easy "composition" of Actions (by instantiating one in the body of another, and delegating `IAction` methods to it).

Thoughts on this approach?

Reply:
I agree. Why not have `IAction` only include a `bind` method which returns an `ActionConfig` with all the details of the action (that's our usual "bind" pattern). Why do we need to separate `bind` and `getActionProperties`?

Reply:
I agree that normally this is the right thing to do. However, in this specific instance, the Pipeline construct needs a few properties of the Action to determine the behavior (`actionName`, `region`, `resource`, etc.). Hence the split between `artifactProperties` (which are more "static") and `ActionConfig` (which contain the Action-specific configuration, and is often more "dynamic").

Reply:
Done in the newest revision.